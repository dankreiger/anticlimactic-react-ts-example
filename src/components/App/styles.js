import styled from 'styled-components';

export const AppContainer = styled.div`
  display: flex;
  flex-direction: column;
  height: 100vh;
  justify-content: center;
  align-items: center;
`;

export const AppInputGroupContainer = styled.div`
  margin-top: 20px;
  margin-bottom: 20px;
`;
