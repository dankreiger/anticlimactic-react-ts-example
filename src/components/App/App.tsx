import * as React from 'react';
import { SFC, useState } from 'react'; // // https://github.com/DefinitelyTyped/DefinitelyTyped/issues/5128
import { IFormInputHook } from 'src/interfaces/hookInterfaces/IFormInputHook.interface';
// hooks
import useDocumentTitle from '../../hooks/useDocumentTitle/useDocumentTitle';
import useFormInput from '../../hooks/useFormInput/useFormInput';
import useWindowWidth from '../../hooks/useWindowWidth/useWindowWidth';

// components
import InputGroup from '../InputGroup/InputGroup';

// styles
import { AppContainer, AppInputGroupContainer } from './styles';

const App: SFC<{}> = () => {
  const name: IFormInputHook = useFormInput('Dan');
  const surname: IFormInputHook = useFormInput('Kreiger');
  useDocumentTitle(`${name.value} ${surname.value}`);

  const [count, setCount] = useState(0);
  const handleClick = () => setCount(count + 1);
  const windowWidth = useWindowWidth();

  return (
    <AppContainer>
      <button onClick={handleClick}>Click me {count}</button>
      <AppInputGroupContainer>
        <InputGroup inputLabel="Name" stateHook={name} />
        <InputGroup inputLabel="Surname" stateHook={surname} />
      </AppInputGroupContainer>
      <p className="windowWidth">Window width: {windowWidth}</p>
    </AppContainer>
  );
};

export default App;
