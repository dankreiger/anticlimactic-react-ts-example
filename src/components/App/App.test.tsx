import { mount, ReactWrapper, render } from 'enzyme';
import * as React from 'react';
import { IGlobal } from 'src/interfaces/testInterfaces/IGlobal.interface';
import App from './App';

declare let global: IGlobal;

describe('App', () => {
  describe('rendering', () => {
    const renderedComponent: Cheerio = render(<App />); // shallow wrapper not working with hooks
    test('renders as expected', () => {
      expect(renderedComponent).toBeTruthy();
      expect(renderedComponent).toMatchSnapshot();
    });
  });

  describe('structure', () => {
    let component: ReactWrapper;
    beforeEach(() => {
      component = mount(<App />); // Enzyme's shallow not working with React state hooks
    });
    describe('button', () => {
      test('renders 1 button html element', () => {
        expect(component.find('button').length).toBe(1);
      });
      test('button has text "Click me"', () => {
        expect(component.find('button').text()).toBe('Click me 0');
      });
    });
    describe('inputs', () => {
      test('renders 2 input html elements', () => {
        expect(component.find('input').length).toBe(2);
      });
    });

    describe('div for window width', () => {
      test('renders 1 div.windowWidth', () => {
        expect(component.find('.windowWidth').length).toBe(1);
      });
      test('div.windowWidth displays the current window width information', () => {
        expect(component.find('.windowWidth').text()).toBe(
          'Window width: 1024'
        ); // 1024 is the JSDOM default
      });
    });
  });

  describe('behavior', () => {
    let component: ReactWrapper;

    test('renders an incremented value on button click', () => {
      component = mount(<App />); // Enzyme's shallow isn't working with state hooks

      expect(component.find('button').text()).toBe('Click me 0');
      component.find('button').simulate('click');
      component.update(); // Read: enzyme update
      expect(component.find('button').text()).toBe('Click me 1');
    });

    test('renders the window size on resize', () => {
      global.innerWidth = 500;
      component = mount(<App />); // only works like this right now - .update() not working on component

      expect(component.find('.windowWidth').text()).toBe('Window width: 500');
      global.innerWidth = 1023;
      component = mount(<App />); // only works like this right now - .update() not working on component
      expect(component.find('.windowWidth').text()).toBe('Window width: 1023');
    });
  });
});
