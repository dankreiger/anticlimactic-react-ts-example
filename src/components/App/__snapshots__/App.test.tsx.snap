// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App rendering renders as expected 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": <Row
        className="align-items-center"
        tag="div"
>
        <Col
                tag="div"
                widths={
                        Array [
                                "xs",
                                "sm",
                                "md",
                                "lg",
                                "xl",
                              ]
                }
        >
                <InputGroup
                        tag="div"
                >
                        <InputGroupAddon
                                addonType="prepend"
                                tag="div"
                        >
                                Name
                        </InputGroupAddon>
                        <Input
                                onChange={[Function]}
                                type="text"
                                value="Dan"
                        />
                </InputGroup>
        </Col>
        <Col
                tag="div"
                widths={
                        Array [
                                "xs",
                                "sm",
                                "md",
                                "lg",
                                "xl",
                              ]
                }
        >
                <InputGroup
                        tag="div"
                >
                        <InputGroupAddon
                                addonType="prepend"
                                tag="div"
                        >
                                Surname
                        </InputGroupAddon>
                        <Input
                                onChange={[Function]}
                                type="text"
                                value="Kreiger"
                        />
                </InputGroup>
        </Col>
        <Col
                tag="div"
                widths={
                        Array [
                                "xs",
                                "sm",
                                "md",
                                "lg",
                                "xl",
                              ]
                }
        >
                <button
                        onClick={[Function]}
                >
                        Click me 
                        0
                </button>
        </Col>
</Row>,
      "tag": "div",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <Col
            tag="div"
            widths={
                        Array [
                                    "xs",
                                    "sm",
                                    "md",
                                    "lg",
                                    "xl",
                                  ]
            }
>
            <InputGroup
                        tag="div"
            >
                        <InputGroupAddon
                                    addonType="prepend"
                                    tag="div"
                        >
                                    Name
                        </InputGroupAddon>
                        <Input
                                    onChange={[Function]}
                                    type="text"
                                    value="Dan"
                        />
            </InputGroup>
</Col>,
          <Col
            tag="div"
            widths={
                        Array [
                                    "xs",
                                    "sm",
                                    "md",
                                    "lg",
                                    "xl",
                                  ]
            }
>
            <InputGroup
                        tag="div"
            >
                        <InputGroupAddon
                                    addonType="prepend"
                                    tag="div"
                        >
                                    Surname
                        </InputGroupAddon>
                        <Input
                                    onChange={[Function]}
                                    type="text"
                                    value="Kreiger"
                        />
            </InputGroup>
</Col>,
          <Col
            tag="div"
            widths={
                        Array [
                                    "xs",
                                    "sm",
                                    "md",
                                    "lg",
                                    "xl",
                                  ]
            }
>
            <button
                        onClick={[Function]}
            >
                        Click me 
                        0
            </button>
</Col>,
        ],
        "className": "align-items-center",
        "tag": "div",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <InputGroup
              tag="div"
>
              <InputGroupAddon
                            addonType="prepend"
                            tag="div"
              >
                            Name
              </InputGroupAddon>
              <Input
                            onChange={[Function]}
                            type="text"
                            value="Dan"
              />
</InputGroup>,
            "tag": "div",
            "widths": Array [
              "xs",
              "sm",
              "md",
              "lg",
              "xl",
            ],
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": Array [
                <InputGroupAddon
                  addonType="prepend"
                  tag="div"
>
                  Name
</InputGroupAddon>,
                <Input
                  onChange={[Function]}
                  type="text"
                  value="Dan"
/>,
              ],
              "tag": "div",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "addonType": "prepend",
                  "children": "Name",
                  "tag": "div",
                },
                "ref": null,
                "rendered": "Name",
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "onChange": [Function],
                  "type": "text",
                  "value": "Dan",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": [Function],
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <InputGroup
              tag="div"
>
              <InputGroupAddon
                            addonType="prepend"
                            tag="div"
              >
                            Surname
              </InputGroupAddon>
              <Input
                            onChange={[Function]}
                            type="text"
                            value="Kreiger"
              />
</InputGroup>,
            "tag": "div",
            "widths": Array [
              "xs",
              "sm",
              "md",
              "lg",
              "xl",
            ],
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": Array [
                <InputGroupAddon
                  addonType="prepend"
                  tag="div"
>
                  Surname
</InputGroupAddon>,
                <Input
                  onChange={[Function]}
                  type="text"
                  value="Kreiger"
/>,
              ],
              "tag": "div",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "addonType": "prepend",
                  "children": "Surname",
                  "tag": "div",
                },
                "ref": null,
                "rendered": "Surname",
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "onChange": [Function],
                  "type": "text",
                  "value": "Kreiger",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": [Function],
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <button
              onClick={[Function]}
>
              Click me 
              0
</button>,
            "tag": "div",
            "widths": Array [
              "xs",
              "sm",
              "md",
              "lg",
              "xl",
            ],
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "Click me ",
                0,
              ],
              "onClick": [Function],
            },
            "ref": null,
            "rendered": Array [
              "Click me ",
              0,
            ],
            "type": "button",
          },
          "type": [Function],
        },
      ],
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <Row
          className="align-items-center"
          tag="div"
>
          <Col
                    tag="div"
                    widths={
                              Array [
                                        "xs",
                                        "sm",
                                        "md",
                                        "lg",
                                        "xl",
                                      ]
                    }
          >
                    <InputGroup
                              tag="div"
                    >
                              <InputGroupAddon
                                        addonType="prepend"
                                        tag="div"
                              >
                                        Name
                              </InputGroupAddon>
                              <Input
                                        onChange={[Function]}
                                        type="text"
                                        value="Dan"
                              />
                    </InputGroup>
          </Col>
          <Col
                    tag="div"
                    widths={
                              Array [
                                        "xs",
                                        "sm",
                                        "md",
                                        "lg",
                                        "xl",
                                      ]
                    }
          >
                    <InputGroup
                              tag="div"
                    >
                              <InputGroupAddon
                                        addonType="prepend"
                                        tag="div"
                              >
                                        Surname
                              </InputGroupAddon>
                              <Input
                                        onChange={[Function]}
                                        type="text"
                                        value="Kreiger"
                              />
                    </InputGroup>
          </Col>
          <Col
                    tag="div"
                    widths={
                              Array [
                                        "xs",
                                        "sm",
                                        "md",
                                        "lg",
                                        "xl",
                                      ]
                    }
          >
                    <button
                              onClick={[Function]}
                    >
                              Click me 
                              0
                    </button>
          </Col>
</Row>,
        "tag": "div",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <Col
              tag="div"
              widths={
                            Array [
                                          "xs",
                                          "sm",
                                          "md",
                                          "lg",
                                          "xl",
                                        ]
              }
>
              <InputGroup
                            tag="div"
              >
                            <InputGroupAddon
                                          addonType="prepend"
                                          tag="div"
                            >
                                          Name
                            </InputGroupAddon>
                            <Input
                                          onChange={[Function]}
                                          type="text"
                                          value="Dan"
                            />
              </InputGroup>
</Col>,
            <Col
              tag="div"
              widths={
                            Array [
                                          "xs",
                                          "sm",
                                          "md",
                                          "lg",
                                          "xl",
                                        ]
              }
>
              <InputGroup
                            tag="div"
              >
                            <InputGroupAddon
                                          addonType="prepend"
                                          tag="div"
                            >
                                          Surname
                            </InputGroupAddon>
                            <Input
                                          onChange={[Function]}
                                          type="text"
                                          value="Kreiger"
                            />
              </InputGroup>
</Col>,
            <Col
              tag="div"
              widths={
                            Array [
                                          "xs",
                                          "sm",
                                          "md",
                                          "lg",
                                          "xl",
                                        ]
              }
>
              <button
                            onClick={[Function]}
              >
                            Click me 
                            0
              </button>
</Col>,
          ],
          "className": "align-items-center",
          "tag": "div",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": <InputGroup
                tag="div"
>
                <InputGroupAddon
                                addonType="prepend"
                                tag="div"
                >
                                Name
                </InputGroupAddon>
                <Input
                                onChange={[Function]}
                                type="text"
                                value="Dan"
                />
</InputGroup>,
              "tag": "div",
              "widths": Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ],
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": Array [
                  <InputGroupAddon
                    addonType="prepend"
                    tag="div"
>
                    Name
</InputGroupAddon>,
                  <Input
                    onChange={[Function]}
                    type="text"
                    value="Dan"
/>,
                ],
                "tag": "div",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "addonType": "prepend",
                    "children": "Name",
                    "tag": "div",
                  },
                  "ref": null,
                  "rendered": "Name",
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "onChange": [Function],
                    "type": "text",
                    "value": "Dan",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
              ],
              "type": [Function],
            },
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": <InputGroup
                tag="div"
>
                <InputGroupAddon
                                addonType="prepend"
                                tag="div"
                >
                                Surname
                </InputGroupAddon>
                <Input
                                onChange={[Function]}
                                type="text"
                                value="Kreiger"
                />
</InputGroup>,
              "tag": "div",
              "widths": Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ],
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": Array [
                  <InputGroupAddon
                    addonType="prepend"
                    tag="div"
>
                    Surname
</InputGroupAddon>,
                  <Input
                    onChange={[Function]}
                    type="text"
                    value="Kreiger"
/>,
                ],
                "tag": "div",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "addonType": "prepend",
                    "children": "Surname",
                    "tag": "div",
                  },
                  "ref": null,
                  "rendered": "Surname",
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "onChange": [Function],
                    "type": "text",
                    "value": "Kreiger",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
              ],
              "type": [Function],
            },
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": <button
                onClick={[Function]}
>
                Click me 
                0
</button>,
              "tag": "div",
              "widths": Array [
                "xs",
                "sm",
                "md",
                "lg",
                "xl",
              ],
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "Click me ",
                  0,
                ],
                "onClick": [Function],
              },
              "ref": null,
              "rendered": Array [
                "Click me ",
                0,
              ],
              "type": "button",
            },
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
